src/wrappers/themis/java/com/cossacklabs/themis/SecureCell.java

```diff
@@ -187,10 +183,10 @@ public SecureCellData protect(byte[] context, byte[] data) throws NullArgumentEx
 	 * @return SecureCellData with protected data
 	 * @throws NullArgumentException when key or data is null
 	 * @throws SecureCellException when cannot protect the data
-	 * @throws UnsupportedEncodingException when UTF-16 decoding is not supported
+	 * @throws RuntimeException when UTF-16 decoding is not supported
 	 */
-	public SecureCellData protect(String password, String context, byte[] data) throws UnsupportedEncodingException, NullArgumentException, SecureCellException {
-		return this.protect(password.getBytes(CHARSET), context.getBytes(CHARSET), data);
+	public SecureCellData protect(String password, String context, byte[] data) throws SecureCellException {
+		return this.protect(Utils.getUTF16Bytes(password), Utils.getUTF16Bytes(context), data);
```

Comment:
why did you decide use a hardcoded method for decoding string instead using constantly configured charset? as I know, java allow to change constants at runtime and if someone will want to change it, it allow to change it from his own code without re-writing themis. But now he will need to inherit a class and re-implement method for it. Better way to refactor `protect` method to be as template method where will be called some other method `decodeString` with default implementation which use `Utils.getUTF16Bytes`, but people will need override only this one

Reply:
oh, Utils it's our new module) so in java way will be better to add dependency injection and use some instance of converter which will be used and which should be provided in new one constructor and used default in our current constructor

Reply:
All places use a hardcoded constant `"UTF-16"`, I’ve only replaced that with a common utility to avoid code duplication and preserve the current behavior and interface.

`CHARSET` constants are declared package-private (= not accessible outside of our package) `final` (= not assignable) `String` (= immutable in Java) values so it’s not really feasible to change them for library users. It’s probably possible with some vile abuse of reflection and unsafe API, but that’s a footgun.

So yeah, there are a number of possible approaches if our users would like to use a different encoding:

- inherit and override all interesting methods individually
- provide a `protected` method for encoding/decoding that can be overriden
- provide a way to inject encoder/decoder into our objects
- provide a way to specify encoding to be passed to `String.getBytes()`
- use composition: wrap Secure Cell into the interface you like
- encode/decode outside of Secure Cell and use `byte[]` API

None of which is probably for us to decide.

I’d personally would like to get rid of String interfaces altogether, unless we agree on the encoding to use across Themis wrappers. Or move it into Secure​Cell​That​Works​Only​With​Java​Strings subclass to keep it cleanly separated from the basic Secure Cell compatible with other platforms. However, that’s a breaking change and it will trigger @vixentael.

My opinion is that Themis works with bytes. If you’d like to use strings that’s great, but please don’t force us to make serialization decisions for you. You’re free to wrap Secure Cell into whatever interface you like and use whatever serialization you like — but it’s there, in your package, not here, in our package.

`com.cossacklabs.themis.SecureCell` makes a serialization decision for the user: `Strings` are serialized as UTF-16. I believe that it’s wrong, given that other platforms use different encodings for their native strings. But that’s how it is right now: out-of-the-box Java strings don’t work with anything other than Java.

Reply:
On the constructive side, while I was typing all of the above I’ve reviewed `String` API and it seems that we don’t need the Utils class. Instead we can replace

```java
password.getBytes(CHARSET)
```

with

```java
password.getBytes(StandardCharsets.UTF_16)
```

which does not throw checked UnsupportedEncodingException because the encoding is known.

Reply:
> I’d personally would like to get rid of String interfaces altogether, unless we agree on the encoding to use across Themis wrappers. Or move it into Secure​Cell​That​Works​Only​With​Java​Strings subclass to keep it cleanly separated from the basic Secure Cell compatible with other platforms. However, that’s a breaking change and it will trigger @vixentael.

While i totally support the unification, Java strings are indeed unique in their `UTF16` usage. 

> current code is not only incompatible with other platforms, it is also incompatible between different Java implementations

luckly, Themis works with bytes and suggest strings API only as data input, not data output

Reply:
Regarding String API deprecation, we may actually deprecate the current API In favor of the proper password-based API when we have it, and this may give us a chance to move the string-based API elsewhere and maybe make it more compatible with other platforms.

Passwords are going to require some decision from us because for passwords it does not make sense to force the users to use `byte[]` for them. We’re likely to enable string usage for passwords for other platforms as well, and here we’ll have issues if some platforms use UTF-16 while others default to UTF-8, or something else (like whatever legacy encoding they are running with; I’m looking at you, Ruby).

> Themis works with bytes and suggest strings API only as data input, not data output

Well, the data is still recoverable, no problem. By “incompatibility” I mean when it does not work successfully our of the box. Of course it will work if the developers use byte-oriented API and do not mess up string encodings then.

However, it may not work in the default use case, when one party uses a `"secret"` in one language, the other uses `"secret"` in their language, but Secure Cell fails to decrypt because Themis wrappers receive different bytes for the master key because the encoding is not the same.

Thankfully, modern languages (finally!) make it painfully obvious that bytes and strings are not the same. You all know the story with Python 3. Swift, for example, also forces the user to make an explicit decision how exactly they turn their `String` into `Data` required by Themis.